src/concurrency.rs

```diff
@@ -0,0 +1,120 @@
+use std::{mem, thread};
+
+use crossbeam_channel::{bounded, Receiver, Sender};
+
+/// `ConcurrentJob` is a handle for some long-running computation
+/// off the main thread. It can be used, indirectly, to wait for
+/// the completion of the said computation.
+///
+/// All `ConncurrentJob`s must eventually be stored in a `Jobs` table.
+///
+/// All concurrent activities, like spawning a thread or pushing
+/// a work item to a job queue, should be covered by `ConcurrentJob`.
+/// This way, the set of `Jobs` table will give a complete overview of
+/// concurrency in the system, and it will be possinle to wait for all
+/// jobs to finish, which helps tremendously with making tests deterministic.
+///
+/// `JobToken` is the worker-side counterpart of `ConcurrentJob`. Dropping
+/// a `JobToken` signals that the corresponding job has finished.
+#[must_use]
+pub struct ConcurrentJob {
+    is_abandoned: bool,
+    chan: Receiver<Never>,
+}
+
+pub struct JobToken {
+    #[allow(unused)] // for drop
+    chan: Sender<Never>,
+}
+
+pub struct Jobs {
+    jobs: Vec<ConcurrentJob>,
+}
+
+impl Jobs {
+    pub fn new() -> Jobs {
+        Jobs { jobs: Vec::new() }
+    }
+
+    pub fn add(&mut self, job: ConcurrentJob) {
+        self.gc();
+        self.jobs.push(job);
+    }
+
+    /// Blocks the current thread until all pending jobs are finished.
+    pub fn wait_for_all(&mut self) {
+        while !self.jobs.is_empty() {
+            let done = {
+                let chans = self.jobs.iter().map(|j| &j.chan);
+                select! {
+                    recv(chans, msg, from) => {
+                        assert!(msg.is_none());
+                        self.jobs.iter().position(|j| &j.chan == from).unwrap()
+                    }
+                }
+            };
+            drop(self.jobs.swap_remove(done));
+        }
+    }
+
+    fn gc(&mut self) {
+        self.jobs.retain(|job| !job.is_completed())
+    }
+}
+
+// For the time being, just detach all currently running
+// jobs on `Drop`. Are more deterministic behavior is
+// to wait for all jobs to finish.
+impl Drop for Jobs {
+    fn drop(&mut self) {
+        let jobs = mem::replace(&mut self.jobs, Vec::new());
+        for job in jobs {
+            job.abandon()
+        }
+    }
+}
+
+impl ConcurrentJob {
+    pub fn new() -> (ConcurrentJob, JobToken) {
+        let (tx, rx) = bounded(0);
+        let job = ConcurrentJob {
+            chan: rx,
+            is_abandoned: false,
+        };
+        let token = JobToken { chan: tx };
+        (job, token)
+    }
+
+    fn is_completed(&self) -> bool {
+        is_closed(&self.chan)
+    }
+
+    fn abandon(mut self) {
+        self.is_abandoned = true
+    }
+}
+
+impl Drop for ConcurrentJob {
+    fn drop(&mut self) {
+        if self.is_abandoned || self.is_completed() || thread::panicking() {
```

Comment:
Why don't we panic on abandoned jobs? Seems like a bad thing if `Jobs` is dropped without completing the jobs

Reply:
Hm, indeed, for normal termination process it is reasonable to require that all the jobs are awaited for!

Still not sure what's the best choice for abnormal termination (some thoughts in af2534f1234da34c8d9bb3d46798196ec95d9d12), but probably just abandoning jobs is fine, as things have gone south anyway.